src/allocation_engine/interval_tree.rs

```diff
@@ -33,14 +37,117 @@ pub struct Constraint {
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord, Debug)]
 pub struct Range {
-    pub min: u64,
-    pub max: u64,
+    /// Lower boundary of the interval.
+    start: u64,
+    /// Upper boundary of the interval.
+    end: u64,
+}
+
+#[allow(clippy::len_without_is_empty)]
+impl Range {
+    /// Create a new Range object.
+    pub fn new(start: u64, end: u64) -> Result<Self> {
+        if start > end {
+            return Err(Error::InvalidRange(start, end));
+        }
+        Ok(Range { start, end })
+    }
+
+    /// Get length of the range.
+    pub fn len(&self) -> u64 {
+        self.end - self.start + 1
+    }
+
+    /// Check whether two Range objects overlap with each other.
+    pub fn overlaps(&self, other: &Range) -> bool {
+        max(self.start, other.start) <= min(self.end, other.end)
+    }
+
+    /// Check whether the key is fully covered.
+    pub fn contains(&self, other: &Range) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+
+    /// Get the beginning of the range.
+    pub fn get_start(&self) -> u64 {
+        self.start
+    }
+
+    /// Get the upper boundary of the range.
+    pub fn get_end(&self) -> u64 {
+        self.end
+    }
+}
+
+/// Returns the first multiple of `alignment` that is lower or equal to the
+/// specified address. This method works only for alignment values that are a
+/// power of two.
+///
+/// # Examples
+/// ```rust
+/// extern crate vm_allocator;
+/// use vm_allocator::allocation_engine::{align_down, Range};
+/// use vm_allocator::Result;
+/// use vm_allocator::Error;
+///
+/// fn intervals_align_down() -> Result<u64> {
+///     let address = 5;
+///     if let Ok(res) = align_down(address, 2) {
+///         if res == 4 {
+///             return Ok(res);
+///         }
+///         return Err(Error::UnalignedAddress);
+///     }
+///     Err(Error::UnalignedAddress)
+/// }
+///
+/// # intervals_align_down().unwrap();
+/// ```
+pub fn align_down(address: u64, alignment: u64) -> Result<u64> {
+    if !alignment.is_power_of_two() {
+        return Err(Error::InvalidAlignment);
+    }
+    Ok(address & !(alignment - 1))
+}
+
+/// Returns the first multiple of `alignment` that is greater or equal to the
+/// specified address. This method works only for alignment values that are a
+/// power of two.
+///
+/// # Examples
+/// ```rust
+/// extern crate vm_allocator;
+/// use vm_allocator::allocation_engine::{align_up, Range};
+/// use vm_allocator::Result;
+/// use vm_allocator::Error;
+///
+/// fn intervals_align_up() -> Result<u64> {
+///     let address = 3;
+///     if let Ok(res) = align_up(address, 2) {
+///         if res == 4 {
+///             return Ok(res);
+///         }
+///         return Err(Error::UnalignedAddress);
+///     }
+///     Err(Error::UnalignedAddress)
+/// }
+///
+/// # intervals_align_up().unwrap();
+/// ```
+pub fn align_up(address: u64, alignment: u64) -> Result<u64> {
+    if alignment == 0 {
```

Comment:
Yes this is correct, but we have to check that alignment is not 0 in this methods as well, because we call `align_down` with `address + (alignment - 1)`

Reply:
Got that ~